operator/cache_memcached.go

```diff
@@ -0,0 +1,222 @@
+package operator
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/bradfitz/gomemcache/memcache"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/grafana/grafana-app-sdk/metrics"
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+var _ cache.Store = &MemcachedStore{}
+
+// MemcachedStore implements cache.Store using memcached as the store for objects.
+// It should be instantiated with NewMemcachedStore.
+type MemcachedStore struct {
+	client       *memcache.Client
+	keyFunc      func(any) (string, error)
+	kind         resource.Kind
+	readLatency  *prometheus.HistogramVec
+	writeLatency *prometheus.HistogramVec
+	keys         sync.Map
```

Comment:
Yeah, that's probably best, because then if the memcached item are lost, the keys will be too, so we don't end up with incompatible states.

Reply:
Although it does mean that adds will need to lock the Store when writing the `keys` key, so adds can only happen sequentially, not concurrently. Not sure if that's a worthwhile trade-off. Thoughts?

Reply:
I think that ends up giving us more possible state errors, such as if the memcached gets cleared (restart or otherwise), if we're using the in-memory key map as the source of truth, we'll be adding keys that don't exist in the cache (and will cause issues when trying to do the cache sync). We can try to handle these cases, by reading the cache keys before write, and then adjusting if the cache keys aren't present, but then that doesn't adjust for other possible state sync errors (though I think this is the biggest one that would be encountered during normal runtime). The in-memory map also has possible state sync errors, so perhaps covering that case with the sync period approach is enough, and we only give a "weak" guarantee of state synchronization with the memcached store, and advise that folks only use it with cache sync if absolutely necessary for their reconciliation process.

Reply:
I'll go forward with the sync interval approach and keep it behind the `TrackKeys` config flag, giving the caveats with `TrackKeys` in the godocs.

Reply:
Nice, thank you!